### DISCUSSION TITLE
Show Dialog When Crashing

### DISCUSSION THREAD

**Discussion by daniel-carr-3000** [AvaloniaUI/Avalonia#16375] [Q&A] [answered]
Hi Everyone! Thanks for your help and input in advance!

I have a small application I've made with Avalonia, I'm attempting to catch any exceptions in place and handle them gracefully, however I've occasionally missed some and my app crashes. I'm currently catching the exceptions globally by doing the following:

```csharp
try
{
    BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);
}
catch (Exception e)
{
    Log.Fatal(e, "An unhandled exception occurred.");
    throw;
}
finally
{
    Log.CloseAndFlush();
}
```

This currently does allow me to get information about the crash by reading my logs, but I would like a way to present a dialog to the user if the app crashes so I can present some information about the exception that occurred and what they might be able to do to get support. 

I haven't been able to figure out a way to handle this. I've tried creating a window in the try / catch above, but Avalonia doesn't like being re-initialized. I've tried catching exceptions in the ```AppDomain.CurrentDomain.UnhandledException``` event, but when I show a window, the process shuts down before you can see and interact with it.

Any suggestions? Thanks!

**Comment by maxkatz6** [comment]
At this line `catch`, Avalonia doesn't run anymore, and you won't be able to show any window.
Ideal solution would be to start another (or the same) app process, and pass error information there. With all additional options like "send error to the developer" if you want.

Alternatively, if you handle Dispatcher.UnhandledException, you still will have some time to show dialog in the same process. But it highly depends on the error - some might still crash the app, and another might cause inconsistent state.

**Reply by maxkatz6** [reply]
You also could move error handling to a more specific location. Like, ICommand handlers. Something what ReactiveUI often does. Making it easier to handle exception closers to the place where they were thrown.

**Reply by daniel-carr-3000** [reply]
I think Dispatcher.UnhandledException is only in the 11.1 beta, but I may try that out. This will really only be for things that slip through so I'll still let the app crash afterwards, I was just hoping to show the user a little more feedback other than the app closing.

**Reply by dbriard** [reply]
Another idea would be to save the error on disk, and present the dialog next time user launch the application.

**Comment by sn4k3** [comment, marked as answer]
I also handle this and present a window to user after the crash:

![image](https://github.com/user-attachments/assets/c604df46-a0c5-4a7f-a7b3-cc6d62118e8e)

All you need is to reopen your program and pass arguments and handle that in startup to present your error.

See: https://github.com/sn4k3/UVtools/blob/master/UVtools.UI/App.axaml.cs#L172
And: https://github.com/sn4k3/UVtools/blob/master/UVtools.UI/Program.cs#L184

**Comment by daniel-carr-3000** [comment]
Thanks guys!
